File: 15_3_sum/solution.ts
function threeSum(nums: number[]): number[][] {
  nums.sort((a, b) => a - b);

  const result: number[][] = [];

  for (let i = 0; i < nums.length - 2; i++) {
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let p1 = i + 1;
    let p2 = nums.length - 1;
    while (p1 < p2) {
      const sum = nums[i] + nums[p1] + nums[p2];

      if (0 < sum) p2--;
      else if (0 > sum) p1++;
      else {
        result.push([nums[i], nums[p1], nums[p2]]);
        p1++;
        while (nums[p1] === nums[p1 - 1] && p1 < p2) {
          p1++;
        }
      }
    }
  }
  return result;
}

File: 15_3_sum/solution.py
from typing import List


class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res: List[List[int]] = []

        for index, num in enumerate(nums):
            if num > 0:
                break

            if index > 0 and num == nums[index - 1]:
                continue

            l, r = index + 1, len(nums) - 1

            while l < r:
                total = num + nums[l] + nums[r]
                if total < 0:
                    l += 1
                elif total > 0:
                    r -= 1
                else:
                    res.append([num, nums[l], nums[r]])
                    l += 1
                    r -= 1
                    while (nums[l] == nums[l - 1]) and l < r:
                        l += 1
        return res


print(Solution().threeSum([-1, -1, 0, 0, 1, 1]))
